### PR TITLE
Python: Add support for handling numpy vectors and JIT compilation with Numba (Closes #34)

### LINKED ISSUE #34: Add support for handling vectorised operations in Python
At the moment we only support scalar inputs, I do not know if there is any demand to support vector operations in other languages but in Python, I think that this would be a good addition.

Having said that, for now, and as a workaround, one can use [numpy.vectorize](https://docs.scipy.org/doc/numpy/reference/generated/numpy.vectorize.html#numpy-vectorize) to allow for vectorised operations in PsychroLib as follows: 

``` py
numpy.vectorize(psychrolib.function_name)
```

E.g.
``` py
import numpy as np
import psychrolib as psy
specific_humidity = [0.0001, 0.0002]
v_GetHumRatioFromSpecificHum = np.vectorize(psy.GetHumRatioFromSpecificHum)
humidity_ratio = v_GetHumRatioFromSpecificHum(specific_humidity)
print(humidity_ratio)
```

``` cmd
array([ 0.00010001,  0.00020004])
```

### PR DESCRIPTION
**Pull request by dmey**
This PR adds the following improvements:

1. Support for handling numpy vectors as inputs/outputs and greater run time performance using Numba: when Numba is found, all functions are compiled using the Numba just in time (JIT) compiler. Note that there is a small overhead the first time the code is run due to the JIT.

2. CI: jobs are now split into 1-2: Python (with/without Numba); 3: C#; 4: R

3. README: content and style.

### CODE REVIEW

**Comment by C-H-Simpson** [comment]
An unfortunate effect of wrapping in this way is that the docstrings are hidden by the wrapper, so e.g. help(psy.GetRelHumFromVapPres) gives "Help on DUFunc in module numba.npyufunc.dufunc object...". There may be a way of fix this, I don't know.

**Comment by dmey** [comment]
> An unfortunate effect of wrapping in this way is that the docstrings are hidden by the wrapper, so e.g. help(psy.GetRelHumFromVapPres) gives "Help on DUFunc in module numba.npyufunc.dufunc object...". There may be a way of fix this, I don't know.

That's right -- [it looks](https://stackoverflow.com/a/29580585) like this will be a limitation so will should probably ask people to just consult the [online API](https://psychrometrics.github.io/psychrolib/api_docs.html) -- not great but given the other options this seems reasonable.

**Comment by dmey** [comment]
@C-H-Simpson how did you handle return type for the Calc functions (e.g. `CalcPsychrometricsFromRelHum`)? I may need to change those for Numba...

**Comment by C-H-Simpson** [comment]
@dmey The Calc... functions can't be straightforwardly vectorized because they return a tuple. 
They should be excluded from the list of functions to wrap with _vectorized_. As the functions it calls to do the calculations are vectorized, it will return a tuple of vectors.
I have tested this and it works.
I don't think there would be much of a speed-up from wrapping the Calc... functions in numba anyway, as they don't actually do any calculations themselves.

So I would recommend just not vectorizing them.

**Comment by dmey** [comment]
@C-H-Simpson thanks -- this makes perfect sense! I will just need to check a few more things/tests so I will need a bit longer before we can release a new version -- hopefully will not be too long though!

**Comment by dmey** [comment]
@didierthevenard this PR is now ready to be merged in -- all tests pass. Please review when you have a moment and let me know for any issue.

File: .travis.yml [hunk 1]
```diff
@@ -1,46 +1,85 @@
...
- on:
-   branch: master
-   condition: $TRAVIS_PYTHON_VERSION == 3.6
+  provider: pages
+  skip_cleanup: true
+  github_token: $GITHUB_TOKEN
+  local_dir: build/html
+  on:
+    branch: master
+    condition: $USE_NUMBA == "no"
```
Comment (didierthevenard) [inline .travis.yml:85]: Add newline at end of file

File: README.md [hunk 2]
```diff
@@ -1,32 +1,46 @@
...
 
 
 ## Documentation
 
-The API documentation is available [here](https://psychrometrics.github.io/psychrolib/api_docs.html). Please note that although the API describes the Python version of the library, the API is common across all the supported language implementations. In R please note that (1) constants, like `ZERO_FAHRENHEIT_AS_RANKINE` are not exported (i.e. not directly accessible to users), (2) functions accept a vector, not a scalar (3) bulk calculations, like `CalcPsychrometricsFromRelHum` return a list.
+Please see the [Python API documentation](https://psychrometrics.github.io/psychrolib/api_docs.html) for the common API across all the supported language implementations. In Python, for best runtime performance and array support, please install [Numba](https://numba.pydata.org/). In R (1) constants, like `ZERO_FAHRENHEIT_AS_RANKINE` are not exported (i.e. not directly accessible to users), (2) functions accept a vector, not a scalar (3) bulk calculations, like `CalcPsychrometricsFromRelHum` return a list.
```
Comment (didierthevenard) [inline README.md:21]: Should mention that installation of numba is optional, that the library stands on its own without numba. Maybe suggest installing numba only for array support.

File: README.md [hunk 3]
```diff
@@ -1,32 +1,46 @@
...
 
 Examples on how to use PsychroLib in all the supported languages are described in [this guide](docs/examples.md).
 
 
-## Installing
+## Installation
 
-- Python: from the [Python Package Index (PyPI)](https://pypi.org/project/PsychroLib/).
+- Python: from the [Python Package Index (PyPI)](https://pypi.org/project/PsychroLib/). For best runtime performance and array support, install [Numba](https://numba.pydata.org/).
```
Comment (didierthevenard) [inline README.md:28]: Same comment as before regarding numba.

File: src/python/psychrolib.py [hunk 4]
```diff
@@ -172,6 +172,13 @@ def SetUnitSystem(Units: UnitSystem) -> None:
     else:
         PSYCHROLIB_TOLERANCE = 0.001
 
+    if has_numba:
```
Comment (didierthevenard) [inline src/python/psychrolib.py:175]: We need comments before this line, to explain to users why this section is needed and what it does.

File: src/python/psychrolib.py [hunk 5]
```diff
@@ -1459,3 +1466,20 @@ def CalcPsychrometricsFromRelHum(TDryBulb: float, RelHum: float, Pressure: float
...
+    isIP = njit(isIP)
+
+    func_list = []
+    for func in list(globals().items()):
+        if isfunction(func[1]) and func[0].startswith(('Get', 'dLnPws_')) and func != 'GetUnitSystem':
+            globals()[func[0]] = vectorize(func[1])
+            func_list.append(func)
+except ImportError:
+    has_numba = False
+    vectorize = None
```
Comment (didierthevenard) [inline src/python/psychrolib.py:1485]: Newline needed at end of file.

File: src/python/psychrolib.py [hunk 6]
```diff
@@ -1459,3 +1466,20 @@ def CalcPsychrometricsFromRelHum(TDryBulb: float, RelHum: float, Pressure: float
     MoistAirVolume = GetMoistAirVolume(TDryBulb, HumRatio, Pressure)
     DegreeOfSaturation = GetDegreeOfSaturation(TDryBulb, HumRatio, Pressure)
     return HumRatio, TWetBulb, TDewPoint, VapPres, MoistAirEnthalpy, MoistAirVolume, DegreeOfSaturation
+
+
```
Comment (didierthevenard) [inline src/python/psychrolib.py:1470]: Comments are needed to explain what this section does (and also why some functions are excluded.

**Comment by dmey** [comment]
> A few comments added to the proposed changes. No issue with the code itself, it's mostly that new code that has been added (around the use of numba and vectorization) needs to have explaining comments.

👍. Please see now.

File: src/python/psychrolib.py [hunk 7]
```diff
@@ -1459,3 +1466,24 @@ def CalcPsychrometricsFromRelHum(TDryBulb: float, RelHum: float, Pressure: float
...
+    # Vectorise all 'core' functions. 
+    # Utility function are excluded as they are just wrappers.
+    func_list = []
+    for func in list(globals().items()):
+        if isfunction(func[1]) and func[0].startswith(('Get', 'dLnPws_')) and func != 'GetUnitSystem':
+            globals()[func[0]] = vectorize(func[1])
+            func_list.append(func)
+except ImportError:
+    has_numba = False
+    vectorize = None
```
Comment (didierthevenard) [inline src/python/psychrolib.py:1489]: What's the purpose of this line? I have the impression that it doesn't do anything.

Comment (dmey) [reply src/python/psychrolib.py]: that's right -- dead code I forgot to remove from the previous version. 👍

**Comment by dmey** [comment]
> Thanks for the added comments. I fixed one typo and added a general comment about the last section in the code (the entire chunk that deals with Numba) - feel free to change if you can think of a better wording. Also wondering about the usefulness of the last line in the code.

Thanks -- looks good. Unused var removed in https://github.com/psychrometrics/psychrolib/pull/57/commits/af6440655f621a38365881ba1f951674d377aa74.

**Pull request merged**